### DISCUSSION TITLE
Select with count option returns count: null but data exists

### DISCUSSION THREAD

**Discussion by NanoCode012** [supabase/supabase#7606] [Questions] [answered]
Hello, I stumbled onto a weird problem. I tried to get the count of a simple table, but received null, while `data` is valid.

I followed https://supabase.com/docs/reference/javascript/select#querying-with-count-option and got the simple below:

```js
const { data, count, error } = await supabase
   .from<definitions["user"]>("user")
    .select("id", { count: "exact" }); // also tried *

if (error) return error;

console.log(error); // null
console.log(count); // null
console.log(data); // Array of Object ids
```

I also tried `planned` and `estimated` and also got `null`. RLS is turned on (see comments for picture), but since data is retrieved successfully, it should not be it.

### Temp workaround 

```js
return data.length ?? 0;
```

## Notes

I found https://github.com/supabase/supabase/discussions/517#discussioncomment-2656471 having the same issue, but there was no continued discussion on it.

If this is a valid error, I could open an Issue on it, but I would like someone to verify the bug exists first if possible.

Env: 
- Typescript 
- "@supabase/supabase-js": "^1.35.4"
- Local development via `supabase start` (Have not tried on supabase's actual db)
- Arkenfox firefox

More info: I am running this within a RTK Query, however, it is not due to cache as I invalidate cache via tags and can see the actual request with correct response via the Network tab.

-------------

EDIT: From discussions with @GaryAustin1 below, we have discovered that hardened firefox (arkenfox) and most likely `1601 Hardened cross origin referers` could be the culprit. I will be switching to chrome for testing for now.

**Comment by silentworks** [comment, marked as answer]
I can't replicate this. Can you share what RLS policies you have in place if any?

Full answer can be found from @GaryAustin1's help below https://github.com/supabase/supabase/discussions/7606#discussioncomment-3115882

**Reply by steve-chavez** [reply]
Hm, even with RLS the count should give 0 and not null. The error must be somewhere else.

**Reply by NanoCode012** [reply]
@silentworks  @silentworks for RLS, I have the basic `Enable select for authenticated users only` 

![image](https://user-images.githubusercontent.com/9899957/178146021-3a38faf5-52ba-4289-853f-a936433b80a3.png)

**Reply by NanoCode012** [reply]
@silentworks , I also checked the network tab for the request. I want to clarify this is with LOCAL development. I have not tried with remote. 

For the GET request, I received status `200 OK` and confirmed that headers contained, `Prefer: count=exact`. Empty request payload. Response payload is JSON array of obj `[{"id":"0e572bfc-037b-4711-93f1-1e396eeef4ae"}]`

If possible, could you share me what is the expected response in the browser's network tab?

**Reply by NanoCode012** [reply]
> Hm, even with RLS the count should give 0 and not null. The error must be somewhere else.

@steve-chavez , I moved the `console.log` above the `if error` and the result is the same.

**Reply by GaryAustin1** [reply]
Do you see the ‘content-range’ header in the response?    
if not is there anything in your stack/browser that could remove it?

**Reply by NanoCode012** [reply]
> Do you see the ‘content-range’ header in the response? if not is there anything in your stack/browser that could remove it?

@GaryAustin1 , thanks for the tip! Your answer made me realize that I hardened my firefox (+CanvasBlocker). I swapped to chrome running incognito and saw that the `count` does work.

For `content-range`: 
- when the table is empty. Firefox(normal) shows : `Content-Range: */0`. 
- when the table has at least 1: Both firefox(normal+incognito) and chrome(incognito) shows `Content-Range: 0-0/1`

Edit: Chrome also sends an extra `OPTIONS` request to the URL `?select=id` before sending a `GET` request to the URL. (Not sure if this helps)

I was wondering if you knew what browser setting could cause this change?

**Reply by GaryAustin1** [reply]
I don't know what things you did to "harden" firefox or if that is some sort of extension.  Basically I'm guessing something is limiting what headers are acceptable to pass.  So if you did anything to white list headers, some extension that deals with headers, cors, etc. would be the likely candidates.

As far as option request, I don't know how that is handled by various browsers for a local server.  For a server on the web I would expect option request with each network request for the most part (there is ability to cache the first option response and not do it on later ones but I've not observed how SB deals with that).

**Reply by NanoCode012** [reply]
@GaryAustin1 , thanks for the reply. Sorry that I was a bit unclear. By hardening firefox, I meant that I used https://github.com/arkenfox/user.js/ .

> if you did anything to white list headers, some extension that deals with headers, cors, etc. would be the likely candidates.

 I will try check those options! I briefly checked the configs, and it could be config `1601 Hardened cross origin referers` (for future readers).

I think I will use Chrome for testing instead then. How should I mark this as resolved? Do you want to create a new top comment that I can mark as resolved or should I resolve the current one?

**Reply by GaryAustin1** [reply]
@silentworks needs some help catching up so checking this one is fine, or add your own if you want to detail the likely culprit in a clean answer.  There have been some other nasty issues lately with odd extensions blocking valid headers, http actions.

**Comment by steve-chavez** [comment]
> const { data, count, error } = await supabase
   .from<definitions["user"]>("user")
    .select("id", { count: "exact" });

Can you remove the typescript annotations and retry the request? Also, does the `user` table exist?

**Reply by NanoCode012** [reply]
@steve-chavez , I have tried without the typescript annotations. Same result.

Yes, the `user` table does exist. The `data` does indeed return the array of `id`s and the `definitions["user"]` does not throw a typescript error, meaning that the table should exist.

**Comment by AbdullahAhmadAAK** [comment]
I'm having the same issue on another table.

```
const { data: previousVisitsCountData, error: previousVisitsCountError } = await supabase
    .from('sessions')
    // .select('*')
    .select('*', { count: 'exact', head: true })
    .eq('ip_address', "192.168.1.10")
```

this also gives me null

**Reply by jonoff** [reply]
You need to use `count` not `data`, as no data is returned when head is set to true. 

I seem to have the reverse issue, probably browser headers again. I am sending a similar query using count but looking at the network log the count/head option is not being passed, as I see full data returned and I get a header response of `content-range: 0-3/*`, and count is set to null.
